src/nonFree/rich_sift/Rich_SIFT_describer.hpp

```diff
@@ -0,0 +1,321 @@
+#include "nonFree/sift/vl/sift.h"
+}
+
+namespace openMVG {
+namespace features {
+
+class Rich_SIFT_Image_describer : public Image_describer
+{
+public:
+
+  using Regions_type = Rich_SIFT_Regions;
+
+  struct Params
+  {
+    Params(
+      int first_octave = 0,
+      int num_octaves = 6,
+      int num_scales = 3,
+      float edge_threshold = 10.0f,
+      float peak_threshold = 0.04f,
+      bool root_sift = true
+    ):
+      _first_octave(first_octave),
+      _num_octaves(num_octaves),
+      _num_scales(num_scales),
+      _edge_threshold(edge_threshold),
+      _peak_threshold(peak_threshold),
+      _root_sift(root_sift) {}
+
+    template<class Archive>
+    inline void serialize( Archive & ar );
+
+    // Parameters
+    int _first_octave;      // Use original image, or perform an upscale if == -1
+    int _num_octaves;       // Max octaves count
+    int _num_scales;        // Scales per octave
+    float _edge_threshold;  // Max ratio of Hessian eigenvalues
+    float _peak_threshold;  // Min contrast
+    bool _root_sift;        // see [1]
+  };
+
+  //--
+  // Constructor
+  //--
+  Rich_SIFT_Image_describer
+  (
+    const Params & params = Params(),
+    bool bOrientation = true
+  ):Image_describer(), _params(params), _bOrientation(bOrientation)
+  {
+    vl_constructor();
+  }
+
+  ~Rich_SIFT_Image_describer()
+  {
+    vl_destructor();
+  }
+
+  bool Set_configuration_preset(EDESCRIBER_PRESET preset) override
+  {
+    switch (preset)
+    {
+    case NORMAL_PRESET:
+      _params._peak_threshold = 0.04f;
+    break;
+    case HIGH_PRESET:
+      _params._peak_threshold = 0.01f;
+    break;
+    case ULTRA_PRESET:
+      _params._peak_threshold = 0.01f;
+      _params._first_octave = -1;
+    break;
+    default:
+      return false;
+    }
+    return true;
+  }
+
+  /**
+  @brief Detect regions on the image and compute their attributes (description)
+  @param image Image.
+  @param regions The detected regions and attributes
+  @param mask 8-bit gray image for keypoint filtering (optional).
+     Non-zero values depict the region of interest.
+  */
+  bool Describe
+  (
+    const image::Image<unsigned char> & image,
+    std::unique_ptr<Regions> &regions,
+    const image::Image<unsigned char> * mask = nullptr
+  ) override
+  {
+    Regions_type *regions_type = static_cast<Regions_type*>(regions.get());
+    if (regions_type == nullptr || regions_type->RegionCount() == 0)
+    {
+      regions = DescribeSIFT(image, mask);
+    }
+    else
+    {
+      DescribeSIFTGivenRegion(image, regions, mask);
+    }
+    return regions != nullptr;
+  }
+
+  /**
+  @brief Detect regions on the image and compute their attributes (description)
+  @param image Image.
+  @param mask 8-bit gray image for keypoint filtering (optional).
+     Non-zero values depict the region of interest.
+  @return regions The detected regions and attributes (the caller must delete the allocated data)
+  */
+  std::unique_ptr<Regions> Describe(
```

Comment:
test this implementation by recomputing all features on image

Reply:
the origin description is calculated using a double orientation, then SIOPointFeature serialize the oritation as a float, so the recalculated description is different, so I convert the orientation to float at the very beginning to solve this.